### PR TITLE
test: [HL-44] Validator unit test

### PR DESCRIPTION
**Pull request by claramounir**
## What type of PR is this? (check all applicable)

- [ ] 🐛 Bug Fix
- [ ] ✅ Test

## Added tests?

- [ ] 👍 yes

## Added to documentation?

- [ ] 🙅 no documentation needed

### CODE REVIEW

**Comment by ahmdmhasn** [comment]
Renamed the title to `test: [HL-44] Validator unit test`

[HL-44]: https://motoon-eg.atlassian.net/browse/HL-44?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by ahmdmhasn** [comment]
@ahmednasr8898 You still need to add the view to the dashboard & attach a screenshot of the result 🫠

File: HealthyTests/Classes/Utilities/Validators/EmailValidatorsTests.swift [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import XCTest
+@testable import Healthy
+final class EmailValidatorsTests: XCTestCase {
+    // MARK: - Properties
```
Comment (mhmdatallaa) [inline HealthyTests/Classes/Utilities/Validators/EmailValidatorsTests.swift:6]:
```suggestion

    // MARK: - Properties
    
```
Follow that pattern in all of your MARKs.

File: HealthyTests/Classes/Utilities/Validators/EmailValidatorsTests.swift [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    var validator: EmailValidator!
+
+    // MARK: - Lifecycle
+    override func setUpWithError() throws {
+        // Put setup code here. This method is called before the invocation of each test method in the class.
+        super.setUp()
+            validator = EmailValidator()
+    }
```
Comment (mhmdatallaa) [inline HealthyTests/Classes/Utilities/Validators/EmailValidatorsTests.swift]:
```suggestion
    override func setUp() {
        super.setUp()
        validator = EmailValidator()
    }
```

File: HealthyTests/Classes/Utilities/Validators/EmailValidatorsTests.swift [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import XCTest
+@testable import Healthy
+final class EmailValidatorsTests: XCTestCase {
```
Comment (mhmdatallaa) [inline HealthyTests/Classes/Utilities/Validators/EmailValidatorsTests.swift]:
```suggestion
import XCTest
@testable import Healthy

final class EmailValidatorsTests: XCTestCase {
```

**Review by mhmdatallaa — COMMENTED** [review]
Great work ya Clara ✨

File: HealthyTests/Classes/Utilities/Validators/EmailValidatorsTests.swift [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        super.setUp()
+            validator = EmailValidator()
+    }
+
+    override func tearDownWithError() throws {
+        // Put teardown code here. This method is called after the invocation of each test method in the class.
+        validator = nil
+            super.tearDown()
+
+    }
```
Comment (mhmdatallaa) [inline HealthyTests/Classes/Utilities/Validators/EmailValidatorsTests.swift]:
```suggestion
```
Not needed as long as you create a new EmailValidator object before each case.

File: HealthyTests/Classes/Utilities/Validators/EmailValidatorsTests.swift [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import XCTest
+@testable import Healthy
+final class EmailValidatorsTests: XCTestCase {
+    // MARK: - Properties
+    var validator: EmailValidator!
```
Comment (mhmdatallaa) [inline HealthyTests/Classes/Utilities/Validators/EmailValidatorsTests.swift]:
```suggestion
    private var validator: EmailValidator!
```

**Review by mhmdatallaa — COMMENTED** [review]
Don't forget to apply the previous changes on `PasswordValidatorTests` too.

File: Healthy/Classes/Utilities/Validators/EmailValidator.swift [hunk 6]
```diff
@@ -3,7 +3,6 @@ import Foundation
 struct EmailValidator: Validator {
     let validationRules: [ValidationRule] = [
         RegexValidationRule(field: L10n.Common.email,
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Validators/EmailValidator.swift:5]: Thanks for fixing this 🙏🏼

Comment (ahmdmhasn) [reply Healthy/Classes/Utilities/Validators/EmailValidator.swift:5]: @esraakhaled FYI

File: HealthyTests/Classes/Utilities/Validators/EmailValidatorsTests.swift [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        let inValidDomain = "test@example.c"
+        XCTAssertFalse(validator.hasValidValue(inValidDomain))
+    }
+    func test_emailValidator_whenEmailStartWithNumber_shouldfailed() {
+        func test_emailValidator_whenEmailHasInValidDomain_shouldfailed() {
+            let startsWithNumber = "123test@example.com"
+            XCTAssertFalse(validator.hasValidValue(startsWithNumber))
+        }
+
+    }
```
Comment (ahmdmhasn) [inline HealthyTests/Classes/Utilities/Validators/EmailValidatorsTests.swift]:
```suggestion
func test_emailValidator_whenEmailIsValid_shouldPass() {
        // Given
        let validEmail = "calara23@gmail.com"

        // Then
        XCTAssertTrue(validator.hasValidValue(validEmail))
    }

    func test_emailValidator_whenEmailIsInValid_shouldFail() {
        // Given
        let noAtSymbolEmail = "testexample.com"

        // Then
        XCTAssertFalse(validator.hasValidValue(noAtSymbolEmail))
    }

    func test_emailValidator_whenEmailDoesNotHaveDomain_shouldFail() {
        // Given
        let noDomain  = "test@"

        // Then
        XCTAssertFalse(validator.hasValidValue(noDomain))
    }

    func test_emailValidator_whenEmailHasInValidDomain_shouldFail() {
        // Given
        let inValidDomain = "test@example.c"

        // Then
        XCTAssertFalse(validator.hasValidValue(inValidDomain))
    }

    func test_emailValidator_whenEmailHasInValidDomain_shouldFail() {
        // Given
        let startsWithNumber = "123test@example.com"

        // Then
        XCTAssertFalse(validator.hasValidValue(startsWithNumber))
    }
```

File: HealthyTests/Classes/Utilities/Validators/PasswordValidatorTests.swift [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    }
+    func test_PasswordValidator_whenPasswordIsNumericOnly_shouldfailed() {
+        let numericPassword = "123456789"
+        XCTAssertFalse(validator.hasValidValue(numericPassword))
+    }
+    func test_PasswordValidator_whenPasswordHasLettersOnly_shouldfailed() {
+        let lettersPassword = "abhjklol;k"
+        XCTAssertFalse(validator.hasValidValue(lettersPassword))
+    }
+}
```
Comment (ahmdmhasn) [inline HealthyTests/Classes/Utilities/Validators/PasswordValidatorTests.swift]:
```suggestion
import XCTest
@testable import Healthy

final class PasswordValidatorTests: XCTestCase {

    // MARK: - Properties

    private var validator: PasswordValidator!

    // MARK: - Lifecycle

    override func setUp() {
        super.setUp()
        validator = PasswordValidator()
    }

    // MARK: - Tests

    func test_PasswordValidator_whenHasValidPassword_shouldPass() {
        // Given
        let validPassword = "SecurePassword123"

        // Then
        XCTAssertTrue(validator.hasValidValue(validPassword))
    }

    func test_PasswordValidator_whenHasShortPassword_shouldFail() {
        // Given
        let shortPassword = "Abc"

        // Then
        XCTAssertFalse(validator.hasValidValue(shortPassword))
    }

    func test_PasswordValidator_whenHasLongPassword_shouldFail() {
        // Given
        let longPassword = "Abc123Abc123Abc123Abc123"

        // Then
        XCTAssertFalse(validator.hasValidValue(longPassword))
    }

    func test_PasswordValidator_whenPasswordIsNumericOnly_shouldFail() {
        // Given
        let numericPassword = "123456789"

        // Then
        XCTAssertFalse(validator.hasValidValue(numericPassword))
    }

    func test_PasswordValidator_whenPasswordHasLettersOnly_shouldFail() {
        // Given
        let lettersPassword = "abhjklol;k"

        // Then
        XCTAssertFalse(validator.hasValidValue(lettersPassword))
    }
}
```

- Spaces updated
- Replace `shouldfailed` with `shouldFail`

**Pull request merged**